service/matching/matching_engine_test.go

```diff
@@ -2499,6 +2497,100 @@ func (s *matchingEngineSuite) TestUnknownBuildId_Match() {
 	wg.Wait()
 }
 
+func (s *matchingEngineSuite) TestUnknownBuildId_Demoted_Match() {
+	namespaceId := namespace.ID(uuid.New())
+	tq := "makeToast"
+	unknown := "unknown"
+	build1 := "build1"
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	s.mockMatchingClient.EXPECT().UpdateWorkerBuildIdCompatibility(gomock.Any(), &matchingservice.UpdateWorkerBuildIdCompatibilityRequest{
+		NamespaceId: namespaceId.String(),
+		TaskQueue:   tq,
+		Operation: &matchingservice.UpdateWorkerBuildIdCompatibilityRequest_PersistUnknownBuildId{
+			PersistUnknownBuildId: unknown,
+		},
+	}).Return(&matchingservice.UpdateWorkerBuildIdCompatibilityResponse{}, nil).AnyTimes() // might get called again on dispatch from spooled
+
+	// add a task for an unknown build id, will get redirected to guessed set
+	_, err := s.matchingEngine.AddWorkflowTask(ctx, &matchingservice.AddWorkflowTaskRequest{
+		NamespaceId:      namespaceId.String(),
+		Execution:        &commonpb.WorkflowExecution{RunId: "run", WorkflowId: "wf"},
+		ScheduledEventId: 123,
+		TaskQueue:        &taskqueuepb.TaskQueue{Name: tq, Kind: enumspb.TASK_QUEUE_KIND_NORMAL},
+		VersionDirective: &taskqueue.TaskVersionDirective{
+			Value: &taskqueue.TaskVersionDirective_BuildId{
+				BuildId: unknown,
+			},
+		},
+	})
+	s.NoError(err)
+	// allow taskReader to finish starting dispatch loop so we don't get an extra load
+	time.Sleep(10 * time.Millisecond)
+
+	// unload base and versioned tqm. note: unload versioned first since versioned taskReader
+	// tries to load base for dispatching.
+	id := newTestTaskQueueID(namespaceId, tq, enumspb.TASK_QUEUE_TYPE_WORKFLOW)
+	verId := newTaskQueueIDWithVersionSet(id, hashBuildId(unknown))
+	verTqm, err := s.matchingEngine.getTaskQueueManager(ctx, verId, normalStickyInfo, false)
+	s.NoError(err)
+	s.NotNil(verTqm)
+	s.matchingEngine.unloadTaskQueue(verTqm)
+	// allow taskReader goroutines time to exit
+	time.Sleep(10 * time.Millisecond)
```

Comment:
Since the taskReader is using goro, can we add a Wait after the Cancel to make this synchronous?

Reply:
good idea, that works here. and it even makes it pass > 99.5% of the time without the first sleep, but I still see some failures there unless I leave the first sleep :(